cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,116 +1273,715 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
     });
 
   // compute total size of each partition
+  // key is the split index
   {
-    // key is split index
     auto keys = cudf::detail::make_counting_transform_iterator(
       0, split_key_functor{static_cast<int>(num_src_bufs)});
     auto values =
       cudf::detail::make_counting_transform_iterator(0, buf_size_functor{d_dst_buf_info});
 
-    thrust::reduce_by_key(rmm::exec_policy(stream),
+    thrust::reduce_by_key(rmm::exec_policy(stream, mr),
```

Comment:
Why adding `mr` here?

Reply:
From the description:

> When thrust and scratch space is required on the GPU, this PR makes use of the memory resource passed, even to the point of using the second argument of exec_policy which is the memory resource. I found this while testing that exec_policy defaults to the per-device resource, and in this case I really wanted to pass a pooled memory resource, outside of our normal async memory resource to set aside this memory ahead of time.

So at this stage in the plugin we are creating a memory resource that is not the async pool for all of this scratch space. It doesn't mean the mr can't be the regular one, but otherwise `rmm::exec_policy` always uses the per-device resource.

Reply:
I'm not quite following this. There are 2 memory resources at play in libcudf. One is the `mr` passed it which is to be only used for allocating returned objects. The default `mr` is to be used for internal temporary memory. If you wan to use a different `mr` for the internal scratch space then can you not just set the default memory resource before calling libcudf?
If there really needs to be a third `mr` here than perhaps we need another parameter to pass that through?
Otherwise, we should at least document that rule breakage here saying were are using the `mr` for the returned data as well as _some_ of the internal scratch space allocations?

Reply:
I made exactly the same comment independently in the header file. Adding a separate `temp_mr` was my proposal.

Reply:
> you want to use a different mr for the internal scratch space then can you not just set the default memory resource before calling libcudf?

No because the default that cuDF is using is the per-device resource (process wide) and Spark is using cuDF from multiple threads. Setting the per-device resource means we would need to lock out all the other threads. A pattern we might want to explore in the future is something like:

```
cudf::call_function_with_default_per_resource_mr();
{
  WITH_SCOPED_MR(my_mr);
  cudf::call_function_with_my_mr();
}
```

E.g. push a mr that isn't the regular mr for a little while.

So just to clarify what is happening in this PR. `cudf::contiguous_split` follows the regular cuDF semantics where there is 1 MR. `cudf::chunked_pack` uses pre-allocated memory and only needs the MR to allocate scratch space and the structures needed to keep the iterative state. I'll move towards `test_mr` for now as it gives us the flexibility to do this, but it is not forced (it would default to the same allocator as `mr`). 

Note that before some of the null count work, we were able to run `chunked_pack` without any allocations on the per-device MR. Now the null count checks happen on the default MR, so this isn't perfect.